src/nebari/plugins.py

```diff
@@ -18,38 +28,243 @@
+            if plugin.endswith(".py"):
+                mod = _import_module_from_filename(plugin)
+            else:
+                mod = importlib.import_module(plugin)
+
+            try:
+                self.plugin_manager.register(mod, plugin)
+            except ValueError:
+                # Pluin already registered
+                pass
+
+    def get_available_stages(self):
+        if not self.exclude_default_stages:
+            self.load_stages(DEFAULT_STAGES_PLUGINS)
+
+        stages = itertools.chain.from_iterable(self.plugin_manager.hook.nebari_stage())
+
+        # order stages by priority
+        sorted_stages = sorted(stages, key=lambda s: s.priority)
+
+        # filter out duplicate stages with same name (keep highest priority)
+        visited_stage_names = set()
+        filtered_stages = []
+        for stage in reversed(sorted_stages):
+            if stage.name in visited_stage_names:
+                continue
+            filtered_stages.insert(0, stage)
+            visited_stage_names.add(stage.name)
+
+        # filter out stages which match excluded stages
+        included_stages = []
+        for stage in filtered_stages:
+            for exclude_stage in self.exclude_stages:
+                if re.fullmatch(exclude_stage, stage.name) is not None:
+                    break
+            else:
+                included_stages.append(stage)
+
+        return included_stages
+
+    def load_config(self, config_path: typing.Union[str, Path]):
+        if isinstance(config_path, str):
+            config_path = Path(config_path)
+
+        if not config_path.exists():
+            raise FileNotFoundError(f"Config file {config_path} not found")
+
+        self.config_path = config_path
+        self.config = schema.read_configuration(config_path)
+
+    def _create_dynamic_schema(
+        self, base: BaseModel, stage: BaseModel, stage_name: str
+    ) -> BaseModel:
+        stage_fields = {
+            n: (f.type_, f.default if f.default is not None else ...)
+            for n, f in stage.__fields__.items()
+        }
+        # ensure top-level key for `stage` is set to `stage_name`
+        stage_model = create_model(stage_name, __base__=schema.Base, **stage_fields)
+        extra_fields = {stage_name: (stage_model, None)}
+        return create_model(self.schema_name, __base__=base, **extra_fields)
+
+    def extend_schema(self, base_schema: BaseModel = schema.Main) -> BaseModel:
+        config_schema = base_schema
+        for stages in self.ordered_stages:
+            if stages.stage_schema:
+                config_schema = self._create_dynamic_schema(
+                    config_schema,
+                    stages.stage_schema,
+                    stages.name,
+                )
+        return config_schema
+
+    def _version_callback(self, value: bool):
+        if value:
+            typer.echo(__version__)
+            raise typer.Exit()
+
+    def create_cli(self) -> typer.Typer:
```

Comment:
I'd prefer if the cli is not coupled to the plugin manager. But I think it makes perfect sense to have a PluginManager.